### PR TITLE
bpo-44524: Fix an issue wherein `_GenericAlias._name` was not properly set for specialforms

### PR DESCRIPTION
**Pull request by BvB93**
An example from prior to this PR:
``` python
>>> from typing import Union
>>> print(Union[int, float].__name__)
None
```

https://bugs.python.org/issue44524

### CODE REVIEW

File: Lib/test/test_typing.py [hunk 1]
```diff
@@ -4793,61 +4793,110 @@ def test_no_isinstance(self):
...
-            typing.MutableSet,
-            typing.OrderedDict,
-            typing.Reversible,
-            typing.Sequence,
-            typing.Set,
-            typing.Sized,
-            typing.Tuple,
-            typing.Type,
-            typing.ValuesView,
+            typing.AbstractSet: 'AbstractSet',
```
Comment (Fidget-Spinner) [inline Lib/test/test_typing.py:4803]: Not sure if we need to be so strict, checking that it's not None/empty string should be enough right?

Then again, it probably won't hurt to have this check in.

File: Misc/NEWS.d/next/Library/2021-08-05-18-20-17.bpo-44524.9T1tfe.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Fixed an issue wherein the `__name__` and `__qualname__` attributes of
+subscribed specialforms could be `None`.
```
Comment (Fidget-Spinner) [inline Misc/NEWS.d/next/Library/2021-08-05-18-20-17.bpo-44524.9T1tfe.rst]: This is making some tests fail at the moment. rST and its double backticks 😉 .
```suggestion
Fixed an issue wherein the ``__name__`` and ``__qualname__`` attributes of
subscribed specialforms could be ``None``.
```

File: Lib/typing.py [hunk 3]
```diff
@@ -1237,6 +1242,12 @@ def __subclasscheck__(self, cls):
             if issubclass(cls, arg):
                 return True
 
+    def __reduce__(self):
+        func, (origin, args) = super().__reduce__()
+        if self._name == "Optional":
+            args = args[1] if args[0] is type(None) else args[0]
+        return func, (origin, args)
```
Comment (Fidget-Spinner) [inline Lib/typing.py]: You can simplify this down a little further:
```suggestion
        func, (origin, args) = super().__reduce__()
        return func, (Union, args)
```

**Review by Fidget-Spinner — COMMENTED** [review]
LGTM in general. Thanks for the quick fix :). I just have a few minor comments below.

**Review by Fidget-Spinner — APPROVED** [review]
Great. Thanks!

**Comment by ambv** [comment]
GHA Windows 32-bit failure in asyncio unrelated.

**Comment by pablogsal** [comment]
Unfortunately this PR has introduced reference leaks:

```
❯ ./python -m test test_typing -R :
0:00:00 load avg: 1.12 Run tests sequentially
0:00:00 load avg: 1.12 [1/1] test_typing
beginning 9 repetitions
123456789
.........
test_typing leaked [29, 29, 29, 29] references, sum=116
test_typing leaked [10, 10, 10, 10] memory blocks, sum=40
test_typing failed (reference leak)

== Tests result: FAILURE ==

1 test failed:
    test_typing

1 re-run test:
    test_typing

Total duration: 1.2 sec
Tests result: FAILURE
```

**Comment by ambv** [comment]
Turns out this PR *discovered* refleaks that were there all along 😎

**Pull request merged**